### PR TITLE
Refactor GestureDetector to call markNeedsSemantics less frequently

### PR DESCRIPTION
**Pull request by goderbauer**
See PR https://github.com/flutter/flutter/pull/11544 for details.

### CODE REVIEW

File: packages/flutter/test/widgets/gesture_detector_test.dart [hunk 1]
```diff
@@ -216,4 +217,89 @@ void main() {
     await tester.tapAt(const Offset(10.0, 10.0));
     expect(didTap, isFalse);
   });
+
+  test('RenderSemanticsGestureHandler.setGestureCallbacks calls markNeedsSemanticsUpdate appropriatly', () {
```
Comment (Hixie) [inline packages/flutter/test/widgets/gesture_detector_test.dart]: appropriately

File: packages/flutter/test/widgets/gesture_detector_test.dart [hunk 2]
```diff
@@ -216,4 +217,89 @@ void main() {
...
+      onLongPress: onLongPress,
+      onHorizontalDragUpdate: onHorizontalDragUpdate,
+      onVerticalDragUpdate: onVerticalDragUpdate,
+    );
+    expect(renderObject.callCount, 4);
+
+    // Change all of them
+    renderObject.setGestureCallbacks(
+      onTap: () {
+        print('tap');
```
Comment (Hixie) [inline packages/flutter/test/widgets/gesture_detector_test.dart]: remove the prints?

**Comment by Hixie** [comment]
Calling markNeedsSemantics redundantly should be a no-op. Can we make markNeedsSemantics more efficient instead?

I'm reluctant to go down the path described in this patch because it breaks with a firmly established pattern in render objects, and would make this render object a special snowflake and thus harder to use (the idea is that people will use these directly when they decide to write their own layer as a replacement for the widgets layer).

**Comment by goderbauer** [comment]
#11544 contains the fix for `markNeedsSemantics` (see PR description) that this behavior uncovered.

I am happy to drop this if this makes the object too much of a special snowflake. It's just an easy optimization I wanted to suggest.

**Pull request closed without merging**